mlflow/server/handlers.py

```diff
@@ -799,6 +823,64 @@ def _delete_model_version_tag():
     return _wrap_response(DeleteModelVersionTag.Response())
 
 
+@catch_mlflow_exception
+def _download_artifact(artifact_path):
+    """
+    A request handler for `GET /mlflow-artifacts/artifacts/<artifact_path>` to download an artifact
+    from `artifact_path` (a relative path from the root artifact directory).
+    """
+    basename = posixpath.basename(artifact_path)
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        tmp_path = os.path.join(tmp_dir, basename)
+        artifact_repo = _get_artifact_repo_mlflow_artifacts()
+        artifact_repo._download_file(artifact_path, tmp_path)
+        return send_from_directory(tmp_dir, basename)
+
+
+@catch_mlflow_exception
+def _upload_artifact(artifact_path):
+    """
+    A request handler for `PUT /mlflow-artifacts/artifacts/<artifact_path>` to upload an artifact
+    to `artifact_path` (a relative path from the root artifact directory).
+    """
+    head, tail = posixpath.split(artifact_path)
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        tmp_path = os.path.join(tmp_dir, tail)
+        with open(tmp_path, "wb") as f:
+            chunk_size = 1024 * 1024  # 1 MB
+            while True:
+                chunk = request.stream.read(chunk_size)
```

Comment:
Is there any timeout on the stream read operation? Else, we may get stuck here forever.

Reply:
- `request.stream` is an instance of `gunicorn.http.body.Body`.
- `gunicorn.http.body.Body.read` doesn't seem to have an option related to `timeout`: 
 
Source: https://github.com/benoitc/gunicorn/blob/ff58e0c6da83d5520916bc4cc109a529258d76e1/gunicorn/http/body.py#L202

```python
    def read(self, size=None):
        size = self.getsize(size)
        ...
```

Reply:
Got it. This probably isn't too big a deal because we set gunicorn worker timeouts. Worst case, we may hang here until the worker times out, which is fine :).